### PR TITLE
Fix formatting for controllers and sigs

### LINKED ISSUE #16956: Indentation errors in rst output from `damlc docs`
# Affected Daml version

2.7.0-snapshot.20230529.11827.0.v3fbe7d01

# Bug description

The latest version of the SDK now shows [the controllers for a given Choice in the docs output](https://github.com/digital-asset/daml/pull/16847). For complex controller lists, the multi-line `rst` output is not properly indented. All lines after the first are dedented by 1 indentation level resulting in `sphinx` throwing warnings while building the docs.

# To reproduce

Minimal reproducible example:

File: Test.daml

```daml
module Test where

data ReceiverAmount = ReceiverAmount
  with
    receiver : Party
    amount : Decimal
  deriving (Show, Eq, Ord)

template Payment
  with 
    sender: Party
    receiverAmounts: [ReceiverAmount]
  where
    signatory sender, map (\r -> r.receiver) receiverAmounts
```

```bash
damlc docs Test.daml -f rst -o docs/
```

Output contains:

```rst
  + **Choice** Archive

    Controller\: sender, map
  (\\ r \-\> (DA\.Internal\.Record\.getField @\"receiver\" r))
  receiverAmounts

    Returns\: ()

    (no fields)
```

# Expected behavior

The indentation should be as follows to prevent sphinx from complaining:

```rst
  + **Choice** Archive

    Controller\: sender, map
    (\\ r \-\> (DA\.Internal\.Record\.getField @\"receiver\" r))
    receiverAmounts

    Returns\: ()

    (no fields)
```

# Additional context

The output for the Signatory section of the template documentation is fine. It seems the bug only affects the Controller section within the documentation for Choices.

### PR DESCRIPTION
**Pull request by samuel-williams-da**
Resolves #16956
May later consider further formatting to split up the line, but I'm quite busy this +next week

~~Adding ghc changes: https://github.com/digital-asset/ghc/pull/168~~ No longer including ghc changes, as it is breaking. Will include for daml3

### CODE REVIEW

**Comment by samuel-williams-da** [comment]
CHeck the new authorities syntax
maybe show authorities?

**Comment by samuel-williams-da** [comment]
Controllers are reversed except if you use {}, due to a mistake (?) in parser.y
Dont reverse in controllers
Can probably fix signatories and braced controllers to not be backwards

File: compiler/damlc/tests/daml-test-files/QualifiedRetroactiveInterfaceInstance.EXPECTED.md [hunk 1]
```diff
@@ -48,7 +48,7 @@
 >
 > * <a name="type-qualifiedretroactiveinterfaceinstance-transfer-84967"></a>**Choice** [Transfer](#type-qualifiedretroactiveinterfaceinstance-transfer-84967)
 >
->   Controller: newOwner, (DA.Internal.Record.getField @"owner" (view this))
+>   Controller: (DA.Internal.Record.getField @"owner" (view this)), newOwner
```
Comment (dylant-da) [inline compiler/damlc/tests/daml-test-files/QualifiedRetroactiveInterfaceInstance.EXPECTED.md:51]: We discussed removing the Internal getField last Monday at the Canton in-person meetup.

- A postprocessing step could detect these expression shapes `...getField @"<fieldName>" <expr>` and fix them back to `<expr>.<fieldName>`
- These blocks *should* have the original source locations embedded - if we could propagate them to this stage, then we could print out what the user originally typed.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract/Templates.hs [hunk 2]
```diff
@@ -200,8 +200,8 @@ isChoiceInst ctx inst
 
   | otherwise = Nothing
 
-hsExprsToString :: [HsExpr GhcPs] -> String
-hsExprsToString = intercalate ", " . fmap (showSDocUnsafe . ppr)
+hsExprToString :: HsExpr GhcPs -> String
+hsExprToString = showSDocOneLine unsafeGlobalDynFlags . ppr
```
Comment (dylant-da) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract/Templates.hs:204]: Moving forward, maybe we should do a review of doc generation and look at places where we could improve formatting. The way we currently pretty print signatories disallows the possibility of a signatory with a multiline definition - we simply compress the expression into a single line, which will probably make readability suffer pretty quickly.

**Review by dylant-da — APPROVED** [review]
Thank you for working on this!
No issues, but I've mentioned some avenues for work for future PRs.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract/Templates.hs (hunk 2 above)
Comment (samuel-williams-da) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract/Templates.hs:204]: Agree, splicing out the source code by location would certainly be nicer

**Pull request merged**